### PR TITLE
Prevent loss of navigation accuracy after in-flight mag yaw alignment

### PR DESCRIPTION
**Pull request by priseborough**
Fixes https://github.com/PX4/ecl/issues/352

See issue for explanation and preliminary test results using replay of FW logs.

The RW testing is still outstanding and will be completed tomorrow.

### CODE REVIEW

**Comment by dagar** [comment]
The CI failures are real.

**Comment by priseborough** [comment]
OK, I have explicitly declared a dynamically sized array which has fixed it.

File: EKF/ekf_helper.cpp [hunk 1]
```diff
@@ -1215,9 +1215,17 @@ void Ekf::zeroCols(float (&cov_mat)[_k_num_states][_k_num_states], uint8_t first
...
 {
-	// store diagonals
+	if (first > last) {
+		return;
+	}
+
+	// create dynamically sized array to save diagonal elemnets
+	const uint8_t array_size = last-first+1;
+	float * variances;
+	variances = new float [array_size];
```
Comment (dagar) [inline EKF/ekf_helper.cpp]: You'd need to delete variances when you're done with it. How about stack allocating `float variances[_k_num_states]` instead? I think a bit of wasted stack is better than a dynamic allocation.

**Comment by priseborough** [comment]
SITL testing of RW flight with a 0.1Ga bias on the X magnetometer axis showed that the practice of using heading fusion to constrain yaw drift while the filed estimates were stabilising made the recovery from mag bias errors slower because the mag heading fusion was 'fighting' the change in yaw from the GPS observations and vehicle motion. This has been addressed by https://github.com/PX4/ecl/pull/354/commits/3f35ad9a76941b238b1d6e44004a7413efa98682

**Comment by priseborough** [comment]
RW testing has been performed with replay of https://logs.px4.io/plot_app?log=78676267-73d6-4588-a4a1-1cfa1dd30af0

Nominal configuration

screen shot 2017-11-10 at 11 26 50 am

Addition of 0.15Ga bias to mag Y axis

screen shot 2017-11-10 at 11 32 07 am

screen shot 2017-11-10 at 11 32 24 am

**Comment by priseborough** [comment]
Tested on a RW vehicle here: https://logs.px4.io/plot_app?log=2b25ce95-28ea-4636-b7b2-db1bd35084f1

**Comment by priseborough** [comment]
Rebased

**Comment by Antiheavy** [comment]
@priseborough we did some fixed wing test flights with this PR rolled in (logs below).  The aircraft flew great on all flights, but looking at the logs there are some EKF Yaw/Airspeed innovation check bits during the flights - anything to do with this PR or do you think it is something else?:
image

2 flights:
https://review.px4.io/plot_app?log=fe5f07bc-93ba-4ca1-a99d-fd3437847e7d

2 flights:
https://review.px4.io/plot_app?log=8757a903-70cd-4d79-9073-1f43707b4a51

**Comment by dagar** [comment]
I took a quick look at the first log and these are airspeed innovation test failures, not yaw.

image

I'll ask if we can get them displayed differently in flight review.
 https://github.com/PX4/flight_review/issues/70

**Comment by Antiheavy** [comment]
Thanks @dagar .  I bet our pitot tube is reading low due to the proximity of the static ports to the leading edge of our wing.  We will gather more data and possibly look into an improved calibration.

**Pull request merged**